L1Trigger/L1THGCal/src/HGCalCoarseTriggerCellMapping.cc

```diff
@@ -0,0 +1,243 @@
+        << "Inconsistent size of coarse trigger cell size vector" << ctcSize_.size();
+  }
+
+  for (auto ctc : ctcSize_)
+    checkSizeValidity(ctc);
+}
+
+const std::map<int, int> HGCalCoarseTriggerCellMapping::kSplit_ = {{kCTCsizeIndividual_, kSplit_v8_Individual_},
+                                                                   {kCTCsizeVeryFine_, kSplit_v8_VeryFine_},
+                                                                   {kCTCsizeFine_, kSplit_v8_Fine_},
+                                                                   {kCTCsizeMid_, kSplit_v8_Mid_},
+                                                                   {kCTCsizeCoarse_, kSplit_v8_Coarse_}};
+
+const std::map<int, int> HGCalCoarseTriggerCellMapping::kSplit_v9_ = {
+    {kCTCsizeIndividual_, kSplit_v9_Individual_},
+    {kCTCsizeVeryFine_, kSplit_v9_VeryFine_},
+    {kCTCsizeFine_, kSplit_v9_Fine_},
+    {kCTCsizeMid_, kSplit_v9_Mid_},
+    {kCTCsizeCoarse_, kSplit_v9_Coarse_},
+};
+
+const std::map<int, int> HGCalCoarseTriggerCellMapping::kSplit_v9_Scin_ = {
+    {kCTCsizeIndividual_, kSplit_v9_Scin_Individual_},
+    {kCTCsizeVeryFine_, kSplit_v9_Scin_VeryFine_},
+    {kCTCsizeFine_, kSplit_v9_Scin_Fine_},
+    {kCTCsizeMid_, kSplit_v9_Scin_Mid_},
+    {kCTCsizeCoarse_, kSplit_v9_Scin_Coarse_},
+};
+
+void HGCalCoarseTriggerCellMapping::checkSizeValidity(int ctcSize) const {
+  if (ctcSize != kCTCsizeFine_ && ctcSize != kCTCsizeCoarse_ && ctcSize != kCTCsizeMid_ &&
+      ctcSize != kCTCsizeVeryFine_ && ctcSize != kCTCsizeIndividual_) {
+    throw cms::Exception("HGCTriggerParameterError")
+        << "Coarse Trigger Cell should be of size " << kCTCsizeIndividual_ << " or " << kCTCsizeVeryFine_ << " or "
+        << kCTCsizeFine_ << " or " << kCTCsizeMid_ << " or " << kCTCsizeCoarse_;
+  }
+}
+
+uint32_t HGCalCoarseTriggerCellMapping::getRepresentativeDetId(uint32_t tcid) const {
+  uint32_t representativeid = 0;
+  DetId tc_Id(tcid);
+  if (tc_Id.det() == DetId::Forward) {  //V8
+    if (triggerTools_.isScintillator(tcid)) {
+      representativeid = tcid;  //stc not available in scintillator for v8
+    } else {
+      representativeid = tcid & ~1;
+    }
+  } else if (tc_Id.det() == DetId::HGCalTrigger || tc_Id.det() == DetId::HGCalHSc) {  //V9
+
+    if (triggerTools_.isScintillator(tcid)) {
+      HGCScintillatorDetId tc_IdV9(tcid);
+      uint32_t newPhi = tc_IdV9.iphi() & ~1;
+      representativeid = tcid & kHGCalScinCellMaskInv_;
+      representativeid |=
+          (((tc_IdV9.ietaAbs() & HGCScintillatorDetId::kHGCalRadiusMask) << HGCScintillatorDetId::kHGCalRadiusOffset) |
+           ((newPhi & HGCScintillatorDetId::kHGCalPhiMask) << HGCScintillatorDetId::kHGCalPhiOffset));
+
+    } else {
+      int uPrime = 0;
+      int newU = 0;
+      int newV = 0;
+      HGCalTriggerDetId tc_IdV9(tcid);
+      int rocnum = detIdToROC_.getROCNumber(tc_IdV9.triggerCellU(), tc_IdV9.triggerCellV(), 1);
+      if (rocnum == 1) {
```

Comment:
should these rocnums be descriptively named constants? this pattern is used in several places